File: app/src/main/java/uk/ivanc/archi/ArchiApplication.java
package uk.ivanc.archi;

import android.app.Application;
import android.content.Context;

import uk.ivanc.archi.model.GithubService;

public class ArchiApplication extends Application {

    private GithubService githubService;

    public static ArchiApplication get(Context context) {
        return (ArchiApplication) context.getApplicationContext();
    }

    public GithubService getGithubService() {
        if (githubService == null) {
            githubService = GithubService.Factory.create();
        }
        return githubService;
    }
}

File: app-2017/src/main/java/uk/ivanc/archi2017/ArchiApplication.kt
package uk.ivanc.archi2017

import android.app.Application
import android.content.Context

import rx.Scheduler
import rx.schedulers.Schedulers
import uk.ivanc.archi2017.model.GithubService

class ArchiApplication : Application() {

    //For setting mocks during testing
    val githubService: GithubService by lazy { GithubService.create() }
    val defaultSubscribeScheduler: Scheduler by lazy { Schedulers.io() }
    val viewModelFactory: ViewModelFactory by lazy { ViewModelFactory(githubService) }

    companion object {

        operator fun get(context: Context): ArchiApplication {
            return context.applicationContext as ArchiApplication
        }
    }
}
